### PR TITLE
move terminal scroll to BufferService - WIP

### LINKED ISSUE #2881: Terminal.scroll should be moved into BufferService
See https://github.com/xtermjs/xterm.js/blob/45077d710627fb2aec61ac2dfc41879d8c32371e/src/browser/Terminal.test.ts#L1421-L1430, https://github.com/xtermjs/xterm.js/pull/2724#pullrequestreview-400997762

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #2881

### CODE REVIEW

File: addons/xterm-addon-webgl/src/WebglRenderer.ts [hunk 1]
```diff
@@ -299,8 +299,8 @@ export class WebglRenderer extends Disposable implements IRenderer {
 
         // Nothing has changed, no updates needed
         if (this._model.cells[i] === code &&
-            this._model.cells[i + RENDER_MODEL_BG_OFFSET] === this._workCell.bg &&
-            this._model.cells[i + RENDER_MODEL_FG_OFFSET] === this._workCell.fg) {
+          this._model.cells[i + RENDER_MODEL_BG_OFFSET] === this._workCell.bg &&
+          this._model.cells[i + RENDER_MODEL_FG_OFFSET] === this._workCell.fg) {
```
Comment (Tyriar) [inline addons/xterm-addon-webgl/src/WebglRenderer.ts:303]: Do you have prettier or something causing these types of changes?

File: src/browser/Terminal.test.ts [hunk 2]
```diff
@@ -117,13 +117,6 @@ describe('Terminal', () => {
       });
       term.resize(1, 1);
     });
-    it('should fire the onScroll event', (done) => {
-      term.onScroll(e => {
-        assert.equal(typeof e, 'number');
-        done();
-      });
-      term.scroll(DEFAULT_ATTR_DATA.clone());
-    });
```
Comment (Tyriar) [inline src/browser/Terminal.test.ts:126]: I think we still want to keep the majority of these tests here, just move the `Terminal InputHandler integration` test suite over.

File: src/common/CoreTerminal.ts [hunk 3]
```diff
@@ -156,133 +171,6 @@ export abstract class CoreTerminal extends Disposable implements ICoreTerminal {
...
-    if (!suppressScrollEvent) {
-      this._onScroll.fire(buffer.ydisp);
-    }
-  }
-
-  /**
-   * Scroll the display of the terminal by a number of pages.
-   * @param pageCount The number of pages to scroll (negative scrolls up).
-   */
-  public scrollPages(pageCount: number): void {
```
Comment (Tyriar) [inline src/common/CoreTerminal.ts:261]: We need to keep all these APIs as they're in xterm.d.ts

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: addons/xterm-addon-webgl/src/WebglRenderer.ts (hunk 1 above)
Comment (meganrogge) [reply addons/xterm-addon-webgl/src/WebglRenderer.ts:303]: seems like it, i'll disable it and revert these

Comment (meganrogge) [reply addons/xterm-addon-webgl/src/WebglRenderer.ts]: I had to turn off format on save.

File: src/common/CoreTerminal.ts [hunk 4]
```diff
@@ -110,6 +110,10 @@ export abstract class CoreTerminal extends Disposable implements ICoreTerminal {
     this.register(forwardEvent(this._coreService.onData, this._onData));
     this.register(forwardEvent(this._coreService.onBinary, this._onBinary));
     this.register(this.optionsService.onOptionChange(key => this._updateOptions(key)));
+    this.register(this._bufferService.onScroll(event => {
+      this._onScroll.fire(event);
```
Comment (Tyriar) [inline src/common/CoreTerminal.ts:114]: onScroll now needs to pass a source when merge conflicts are resolved, see https://github.com/xtermjs/xterm.js/pull/3205

File: src/browser/services/RenderService.ts [hunk 5]
```diff
@@ -72,6 +72,8 @@ export class RenderService extends Disposable implements IRenderService {
     // No need to register this as renderer is explicitly disposed in RenderService.dispose
     this._renderer.onRequestRedraw(e => this.refreshRows(e.start, e.end, true));
 
+    this._renderer.onRecoverContext(() => this.recoverContext());
```
Comment (Tyriar) [inline src/browser/services/RenderService.ts:75]: recover context stuff somehow got mixed in with this change

**Review by Tyriar — CHANGES_REQUESTED** [review]

**Comment by meganrogge** [comment]
closing in favor of #3277

**Pull request closed without merging**